packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,364 @@
+import { Stream } from "@libp2p/interface-connection";
+import type { Libp2p } from "@libp2p/interface-libp2p";
+import type { PeerId } from "@libp2p/interface-peer-id";
+import type { Peer } from "@libp2p/interface-peer-store";
+import type { IncomingStreamData } from "@libp2p/interface-registrar";
+import type {
+  Callback,
+  ContentTopic,
+  IAsyncIterator,
+  IDecodedMessage,
+  IDecoder,
+  IFilterV2,
+  IProtoMessage,
+  IReceiver,
+  PeerIdStr,
+  PeerSubscription,
+  ProtocolCreateOptions,
+  ProtocolOptions,
+  PubSubTopic,
+  SubscriptionsLog,
+  Unsubscribe,
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import { groupByContentTopic, toAsyncIterator } from "@waku/utils";
+import debug from "debug";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { BaseProtocol } from "../../base_protocol.js";
+import { DefaultPubSubTopic } from "../../constants.js";
+
+import {
+  FilterPushRpc,
+  FilterSubscribeResponse,
+  FilterSubscribeRpc,
+} from "./filter_rpc.js";
+
+const log = debug("waku:filter:v2");
+
+const FilterV2Codecs = {
+  SUBSCRIBE: "/vac/waku/filter-subscribe/2.0.0-beta1",
+  PUSH: "/vac/waku/filter-push/2.0.0-beta1",
+};
+
+class Subscription {
+  private readonly peer: Peer;
+  private readonly pubSubTopic: PubSubTopic;
+  private newStream: (peer: Peer) => Promise<Stream>;
+
+  static instances = new Map<[PubSubTopic, PeerIdStr], Subscription>();
```

Comment:
I would try to avoid using `static` variable except when needed.
In this instance, it is not needed. The `FilterV2` class should keep track of the subscriptions.

Then, each `Subscription` instance should keep track of the callback functions.

See https://github.com/waku-org/js-waku/pull/1371